File: c++/advanced/4a_valid_palindrome.cpp
private:
    bool isAlphanumeric(char c){
        if((c >= 'a' && c<='z') || (c>='A' && c<='Z') || (c>='0' && c<='9')){
            return true;
        }
        return false;
    }
    char caseChange(char c){
            char s = tolower(c);
            return s;
        
    }
    void reverse(string &f, string &rev, int i){
        //BC
        if(i>=f.size()) return;
        //Store current char
        char ch = f[i];
        //RC
        i++;
        reverse(f, rev, i);
        //
        rev.push_back(ch);
    }
public:
    bool isPalindrome(string s) {
        string f = "";
        for(auto i: s){
            if(isAlphanumeric(i)){
                if(isupper(i)){
                    f.push_back(caseChange(i));
                }
                else{
                    f.push_back(i);
                }
            }
        }
        string rev = "";
        reverse(f, rev, 0);
    
        if(f == rev) return true;
        else return false;
    }
};

File: js/advanced/4a_valid_palindrome.js


var isPalindrome = function(s) {

    const n = s.length
    let l = 0, r = n-1
    let st = s.toLowerCase()
    const isAlphanumeric = char => /^[a-zA-Z0-9]$/.test(char)
    while(l<=r){
        while(l<n && !isAlphanumeric(s[l])) l++
        while(r>=0 && !isAlphanumeric(s[r])) r--
        if(st[l] !== st[r]) return false
        l++
        r--
    }
    return true
};